feat(ars548): initial driver

knzo25 opened the pull request:
## PR Type

<!-- Select one and remove others. If an appropriate one is not listed, please write by yourself. -->

- New Feature

## Related Links

<!-- Please write related links to GitHub/Jira/Slack/etc. -->

## Description
Adds the initial hardware interface, decoder and ros2 wrapper for the continental ARS548 radar

<!-- Describe what this PR changes. -->

## Review Procedure

<!-- Explain how to review this PR. -->

## Remarks

<!-- Write remarks as you like if you need them. -->

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] (Optional) Unit tests have been written for new behavior
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

knzo25 commented:
@mojomex 

Last time you reviewed until [2338af7](https://github.com/tier4/nebula/pull/123/commits/2338af727c305890391d6daa356c55319e204dc5)
@drwnz has screenshots of the comments from the previous PR in case you want to see them

The radars are already being used here and there, together with other sensors not in the main branch, so I think we need to merge at least this one rather fast (we can always patch things later)

mojomex commented on nebula_ros/launch/continental_launch_all_hw.xml:
Are the "new_" arguments called that for compatibility reasons (to not clash with existig names) or for clarification (as in, these are the new values that nebula will send to the sensor on startup)?

knzo25 replied:
The "new_" are the ones that nebula will try to set the devices to.
I know that for other hardware interfaces this is done automatically, but it is not done at startup and it rather requires us to ask for it via  a service. 

The reasons for this are:
 - No warranty that we will bot brick the devices
 - Some parameters will not take effect until the next power-up
 - Previous devices (or continental drivers I can not remember) came with the warning of discouraging users of unnecessary writes to the internal memory

mojomex replied:
I see! The cleanest approach would be to send these parameters in the service request. The way it is now, it might be confusing for people expecting it to be set automatically.

If changing it to the suggested approach is too cumbersome, documenting the behavior in the launch file and readme would be acceptable too.

knzo25 replied:
Ok, I will make a service for the radar parameters and the sensor ip.
However, the mounting and vehicle parameters need to be ros parameters, as they are normal ros concepts (for example, the pointcloud container computes the vehicle dimensions, and the robot model the /tf)

knzo25 replied:
@mojomex 
The changed were added in 
c7937d5

mojomex commented on nebula_ros/src/continental/continental_ars548_decoder_ros_wrapper.cpp:
```diff
@@ -0,0 +1,851 @@
+      track_msg.classification = PEDESTRIAN_ID;
+    }
+
+    track_msg.position_covariance[0] = static_cast<float>(object.position_std.x);
+    track_msg.position_covariance[1] = object.position_covariance_xy;
+    track_msg.position_covariance[2] = 0.f;
+    track_msg.position_covariance[3] = static_cast<float>(object.position_std.y);
+    track_msg.position_covariance[4] = 0.f;
+    track_msg.position_covariance[5] = static_cast<float>(object.position_std.z);
+
+    track_msg.velocity_covariance[0] = static_cast<float>(object.absolute_velocity_std.x);
+    track_msg.velocity_covariance[1] = object.absolute_velocity_covariance_xy;
+    track_msg.velocity_covariance[2] = 0.f;
+    track_msg.velocity_covariance[3] = static_cast<float>(object.absolute_velocity_std.y);
+    track_msg.velocity_covariance[4] = 0.f;
+    track_msg.velocity_covariance[5] = static_cast<float>(object.absolute_velocity_std.z);
+
+    track_msg.acceleration_covariance[0] = static_cast<float>(object.absolute_acceleration_std.x);
+    track_msg.acceleration_covariance[1] = object.absolute_acceleration_covariance_xy;
+    track_msg.acceleration_covariance[2] = 0.f;
+    track_msg.acceleration_covariance[3] = static_cast<float>(object.absolute_acceleration_std.y);
+    track_msg.acceleration_covariance[4] = 0.f;
+    track_msg.acceleration_covariance[5] = static_cast<float>(object.absolute_acceleration_std.z);
```
Going by the variable names `_std` etc. you are populating the covariance matrix with standard deviations in place of variances, is this intentional?

mojomex commented on nebula_ros/src/continental/continental_ars548_decoder_ros_wrapper.cpp:
```diff
@@ -0,0 +1,851 @@
+
+    const double half_length = 0.5 * object.shape_length_edge_mean;
+    const double half_width = 0.5 * object.shape_width_edge_mean;
+    const int reference_index = std::min<int>(object.position_reference, 8);
```
A comment like "position references 0-7 are valid. In case of invalid references, use the dummy reference at index 8" would make this more clear imo!

knzo25 replied:
Added documentation in [ec80eab](https://github.com/tier4/nebula/pull/123/commits/ec80eab73f575aad21c9521e0db734a0474ef639)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+  const uint32_t number_of_detections = detection_list.number_of_detections.value();
+  msg.detections.resize(number_of_detections);
+
+  if (radar_status_.timestamp_sync_status == "SYNC_OK") {
```
This would be clearer with a comment like "Estimate rate of dropped detections"

knzo25 replied:
Added a comment in [da81bd6](https://github.com/tier4/nebula/pull/123/commits/da81bd619f2f346d53cd6daedc885fc4929ab4f3)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+
+  msg.objects.resize(number_of_objects);
+
+  if (radar_status_.timestamp_sync_status == "SYNC_OK") {
```
This would be clearer with a comment like "Estimate rate of dropped objects"

knzo25 replied:
Added a comment in [da81bd6](https://github.com/tier4/nebula/pull/123/commits/da81bd619f2f346d53cd6daedc885fc4929ab4f3)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+  radar_status_.timestamp_nanoseconds = sensor_status_packet.stamp.timestamp_nanoseconds.value();
+  radar_status_.timestamp_seconds = sensor_status_packet.stamp.timestamp_seconds.value();
+
+  if (sensor_status_packet.stamp.timestamp_sync_status == 1) {
+    radar_status_.timestamp_sync_status = "SYNC_OK";
+  } else if (sensor_status_packet.stamp.timestamp_sync_status == 2) {
+    radar_status_.timestamp_sync_status = "NEVER_SYNC";
+  } else if (sensor_status_packet.stamp.timestamp_sync_status == 3) {
+    radar_status_.timestamp_sync_status = "SYNC_LOST";
+  } else {
+    radar_status_.timestamp_sync_status = "INVALID_VALUE";
+  }
```
This would be cleaner as a switch statement

knzo25 replied:
Addressed in [800b351](https://github.com/tier4/nebula/pull/123/commits/800b351ad5c941b14e31507c4cd59eb6b42ddd4a)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+  radar_status_.sw_version_minor = sensor_status_packet.sw_version_minor;
+  radar_status_.sw_version_patch = sensor_status_packet.sw_version_patch;
+
+  radar_status_.plug_orientation = sensor_status_packet.status.plug_orientation == 0 ? "PLUG_RIGHT"
+                                   : sensor_status_packet.status.plug_orientation == 1
+                                     ? "PLUG_LEFT"
+                                     : "INVALID_VALUE";
```
This would be cleaner as a switch statement

knzo25 replied:
Addressed in [800b351](https://github.com/tier4/nebula/pull/123/commits/800b351ad5c941b14e31507c4cd59eb6b42ddd4a)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+  radar_status_.height = sensor_status_packet.status.height.value();
+  radar_status_.wheel_base = sensor_status_packet.status.wheelbase.value();
+
+  if (sensor_status_packet.status.frequency_slot == 0) {
+    radar_status_.frequency_slot = "0:Low (76.23 GHz)";
+  } else if (sensor_status_packet.status.frequency_slot == 1) {
+    radar_status_.frequency_slot = "1:Mid (76.48 GHz)";
+  } else if (sensor_status_packet.status.frequency_slot == 2) {
+    radar_status_.frequency_slot = "2:High (76.73 GHz)";
+  } else {
+    radar_status_.frequency_slot = "INVALID VALUE";
+  }
```
This would be cleaner as a switch statement

knzo25 replied:
Addressed in [800b351](https://github.com/tier4/nebula/pull/123/commits/800b351ad5c941b14e31507c4cd59eb6b42ddd4a)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+  radar_status_.cycle_time = sensor_status_packet.status.cycle_time;
+  radar_status_.time_slot = sensor_status_packet.status.time_slot;
+
+  radar_status_.hcc = sensor_status_packet.status.hcc == 1 ? "Worldwide"
+                      : sensor_status_packet.status.hcc == 2
+                        ? "Japan"
+                        : ("INVALID VALUE=" + std::to_string(sensor_status_packet.status.hcc));
```
This would be cleaner as a switch statement

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+                        ? "Japan"
+                        : ("INVALID VALUE=" + std::to_string(sensor_status_packet.status.hcc));
+
+  radar_status_.power_save_standstill =
+    sensor_status_packet.status.powersave_standstill == 0   ? "Off"
+    : sensor_status_packet.status.powersave_standstill == 1 ? "On"
+                                                            : "INVALID VALUE";
```
This would be cleaner as a switch statement

knzo25 replied:
Addressed in [800b351](https://github.com/tier4/nebula/pull/123/commits/800b351ad5c941b14e31507c4cd59eb6b42ddd4a)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+    radar_status_.radar_status = "STATE_INIT";
+  } else if (sensor_status_packet.radar_status == 1) {
+    radar_status_.radar_status = "STATE_OK";
+  } else if (sensor_status_packet.radar_status == 2) {
+    radar_status_.radar_status = "STATE_INVALID";
+  } else {
+    radar_status_.radar_status = "INVALID VALUE";
+  }
+
+  if (sensor_status_packet.voltage_status == 0) {
+    radar_status_.voltage_status = "Ok";
+  }
+  if (sensor_status_packet.voltage_status & 0x01) {
+    radar_status_.voltage_status += "Current under voltage";
+  }
+  if (sensor_status_packet.voltage_status & 0x02) {
+    radar_status_.voltage_status = "Past under voltage";
+  }
+  if (sensor_status_packet.voltage_status & 0x03) {
+    radar_status_.voltage_status = "Current over voltage";
+  }
+  if (sensor_status_packet.voltage_status & 0x04) {
+    radar_status_.voltage_status = "Past over voltage";
+  }
+
+  if (sensor_status_packet.temperature_status == 0) {
+    radar_status_.temperature_status = "Ok";
+  }
+  if (sensor_status_packet.temperature_status & 0x01) {
+    radar_status_.temperature_status += "Current under temperature";
+  }
+  if (sensor_status_packet.temperature_status & 0x02) {
+    radar_status_.temperature_status += "Past under temperature";
+  }
+  if (sensor_status_packet.temperature_status & 0x03) {
+    radar_status_.temperature_status += "Current over temperature";
+  }
+  if (sensor_status_packet.temperature_status & 0x04) {
+    radar_status_.temperature_status += "Past over temperature";
+  }
```
These would be cleaner as switch statements

mojomex replied:
* The values `0x03, 0x04` are incorrect and should be `0x04, 0x08` respectively (bitfield, see radar IO datasheet). 
* For voltage, there can also be any combination of the four error messages and they should thus be concatenated.
* Overvoltage etc. should be spelled as one word
* When concatenating, separating the entries with commas would be nice. This is a pain in CPP but there is a Boost function: `boost::algorithm::join(..)`. Alternatively, trailing commas would also be acceptable imo

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+                                                             : "INVALID VALUE";
+
+  radar_status_.longitudinal_acceleration_status =
+    sensor_status_packet.longitudinal_acceleration_status == 0   ? "VDY_OK"
+    : sensor_status_packet.longitudinal_acceleration_status == 1 ? "VDY_NOTOK"
+                                                                 : "INVALID VALUE";
+
+  radar_status_.lateral_acceleration_status =
+    sensor_status_packet.lateral_acceleration_status == 0   ? "VDY_OK"
+    : sensor_status_packet.lateral_acceleration_status == 1 ? "VDY_NOTOK"
+                                                            : "INVALID VALUE";
+
+  radar_status_.yaw_rate_status = sensor_status_packet.yaw_rate_status == 0   ? "VDY_OK"
+                                  : sensor_status_packet.yaw_rate_status == 1 ? "VDY_NOTOK"
+                                                                              : "INVALID VALUE";
+
+  radar_status_.steering_angle_status = sensor_status_packet.steering_angle_status == 0 ? "VDY_OK"
+                                        : sensor_status_packet.steering_angle_status == 1
+                                          ? "VDY_NOTOK"
+                                          : "INVALID VALUE";
+
+  radar_status_.driving_direction_status =
+    sensor_status_packet.driving_direction_status == 0   ? "VDY_OK"
+    : sensor_status_packet.driving_direction_status == 1 ? "VDY_NOTOK"
+                                                         : "INVALID VALUE";
+
+  radar_status_.characteristic_speed_status =
+    sensor_status_packet.characteristic_speed_status == 0   ? "VDY_OK"
+    : sensor_status_packet.characteristic_speed_status == 1 ? "VDY_NOTOK"
+                                                            : "INVALID VALUE";
+
+  if (sensor_status_packet.radar_status == 0) {
+    radar_status_.radar_status = "STATE_INIT";
+  } else if (sensor_status_packet.radar_status == 1) {
+    radar_status_.radar_status = "STATE_OK";
+  } else if (sensor_status_packet.radar_status == 2) {
+    radar_status_.radar_status = "STATE_INVALID";
+  } else {
+    radar_status_.radar_status = "INVALID VALUE";
+  }
```
These would be cleaner as switch statements

knzo25 replied:
Addressed in [800b351](https://github.com/tier4/nebula/pull/123/commits/800b351ad5c941b14e31507c4cd59eb6b42ddd4a)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+  const uint8_t & blockage_status0 = sensor_status_packet.blockage_status & 0x0f;
+  const uint8_t & blockage_status1 = (sensor_status_packet.blockage_status & 0xf0) >> 4;
+
+  if (blockage_status0 == 0) {
+    radar_status_.blockage_status = "Blind";
+  } else if (blockage_status0 == 1) {
+    radar_status_.blockage_status = "High";
+  } else if (blockage_status0 == 2) {
+    radar_status_.blockage_status = "Mid";
+  } else if (blockage_status0 == 3) {
+    radar_status_.blockage_status = "Low";
+  } else if (blockage_status0 == 4) {
+    radar_status_.blockage_status = "None";
+  } else {
+    radar_status_.blockage_status = "INVALID VALUE";
+  }
+
+  if (blockage_status1 == 0) {
+    radar_status_.blockage_status += ". Self test failed";
+  } else if (blockage_status1 == 1) {
+    radar_status_.blockage_status += ". Self test passed";
+  } else if (blockage_status1 == 2) {
+    radar_status_.blockage_status += ". Self test ongoing";
+  } else {
+    radar_status_.blockage_status += ". Invalid self test";
+  }
```
These would be clearer as switch statements

knzo25 replied:
Addressed in [800b351](https://github.com/tier4/nebula/pull/123/commits/800b351ad5c941b14e31507c4cd59eb6b42ddd4a)

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,574 @@
+    radar_status_.radar_status = "INVALID VALUE";
+  }
+
+  if (sensor_status_packet.voltage_status == 0) {
+    radar_status_.voltage_status = "Ok";
+  }
+  if (sensor_status_packet.voltage_status & 0x01) {
+    radar_status_.voltage_status += "Current under voltage";
+  }
+  if (sensor_status_packet.voltage_status & 0x02) {
+    radar_status_.voltage_status = "Past under voltage";
+  }
+  if (sensor_status_packet.voltage_status & 0x03) {
+    radar_status_.voltage_status = "Current over voltage";
+  }
+  if (sensor_status_packet.voltage_status & 0x04) {
+    radar_status_.voltage_status = "Past over voltage";
+  }
```
The bitfields for voltage/temperature status are currently not parsed correctly (0x03, 0x04 are not correct for overvoltage/overtemperature; there can be combinations of current/past over/undervoltage).
It should be `0x04, 0x08` respectively.
Also, undervoltage etc. should be spelled as single words.

It would also be nice to have the statuses comma separated in case a combination of them applies but I realize this is a pain in CPP, so I think comma separation with a trailing comma is also fine (or using `boost::algorithm::join(..)`).

(see IO datasheet)

mojomex commented on nebula_ros/src/continental/multi_continental_ars548_hw_interface_ros_wrapper.cpp:
What is the motivation for the `multi_continental_ars548` driver? For multiple LiDARs, we launch multiple nebula instances because the decoding of each sensor is independent.
As far as I can tell, this is the case for radar as well. Given that the `multi_` approach here would need to be implemented for every radar model, I would suggest removing it and leaving the multi-sensor handling to the module launching nebula.

knzo25 replied:
The motivation for the multi hardware interface for the ARS548 is that the sensors use a fixed multicasting ip, meaning that if we launch 6 hardware interfaces, all hardware interfaces receive the packages from all 6 lidars (we can filter by sender IP, which we do).
To avoid that issue I made a single multi radar hardware interface, which received all packets, and publishes them to the respective topics (there are multiple decoders).

In the experiment were we used 6x ARS548 we had data issues, so I put this up just to check if things got better. I spoke with David, and I know he is against the multi hardware interface, but I would like to keep it until we finish all evaluations (if by then there are no issues, I will send a PR deleting it)

mojomex replied:
I see, fair enough. Could you add a comment to the top of the relevant files saying that it's for evaluation purposes? Just so the next person coming along can understand why it's there :pray:

knzo25 replied:
Added the comment in [f1e1779](https://github.com/tier4/nebula/pull/123/commits/f1e177964608cda122329e2926b5fb6cc6732b2c)

mojomex reviewed (changes requested):
Hi! I have reviewed the new version and see that the previously undecided functionality is now added :+1: 
Apart from the comments above, this looks really good already, thanks!

knzo25 commented:
@mojomex 
Thanks for handling the review so fast. I just pushed the changes addressing your reviews :pray:

amc-nu commented:
I forgot to mention. Can you please add the tests related to the new sensor decoder?

cc @drwnz

knzo25 commented:
@amc-nu 
Thanks for the comments

First, regarding `EthernetSensor` and `LidarSensor` and so on, the reasoning is that in the next PR we will introduce CAN-based sensors, so I prepared the OOP structure to make things fall into place before handed (Already talked with David about this one).

Regarding the second part, after applying the changes corresponding to the `EthernetSensor` and so on, I had to (just slightly) modify several classes. However, as you probably know, pre-commit is not passing on any of the older files, so I just accepted or complied with all the `pre-commit` "suggestions". I will cherry pick the changes to the "unrelated classes", but sooner than later we will need to make the code comply.

As for the tests, noted. I will write them soon enough

amc-nu commented:
@knzo25 Thank you, I confirmed them. Looks great!
Finally, I see you still need to modify the README file.
Can you please explain how to use the driver and launch the radar(s) in both online and offline modes (if any specific parameter is required)?

knzo25 commented:
Thanks for the catch. Implemented the changes to the python script so we can also launch the radar (with online & offline modes)
Also updated the supported models in the readme. As it is growing longer and difficult to read for the first time, David was considering redoing it during his nebula sprint in the next months, so more detailed information will be written then

```
ros2 launch nebula_ros nebula_launch.py sensor_model:=ARS548
```

mojomex commented on nebula_ros/include/nebula_ros/continental/multi_continental_ars548_hw_interface_ros_wrapper.hpp:
```diff
@@ -0,0 +1,146 @@
+}
+
+/// @brief Hardware interface ros wrapper of continental radar ethernet driver
+/// NOTE: this is a temporal class that acts as a single hw interface for multiple devices
```
(pedantic) temporary

mojomex replied:
Addressed in [764e584](https://github.com/tier4/nebula/pull/123/commits/764e584d1e9caa722e73307123f48223a51e265a)

mojomex commented on nebula_ros/config/continental/SRR520.yaml:
Please move this to #124 if possible :pray:

knzo25 replied:
Addressed in [1007571](https://github.com/tier4/nebula/pull/123/commits/100757164b9eee0ced387ed2c5276687796abbfd)

mojomex commented on nebula_ros/launch/continental_launch_all_hw.xml:
```diff
@@ -0,0 +1,174 @@
+            <param name="receiver_timeout_sec" value="$(var receiver_timeout_sec)"/>
+            <param name="sender_timeout_sec" value="$(var sender_timeout_sec)"/>
+            <param name="filters" value="$(var filters)"/>
+            <param name="use_bus_time" value="$(var use_bus_time)"/>
+
+            <param name="new_longitudinal_cog" value="1.25"/> <!-- test value -->
+            <param name="new_sensor_id" value="$(var new_sensor_id)"/>
+            <param name="new_vehicle_wheelbase" value="$(var new_vehicle_wheelbase)"/>
+            <param name="new_plug_bottom" value="$(var new_plug_bottom)"/>
+            <param name="new_cover_damping" value="$(var new_cover_damping)"/>
+            <param name="reset_sensor_configuration" value="$(var reset_sensor_configuration)"/>
+        </node>
+
+        <node pkg="nebula_ros" exec="continental_srr520_hw_interface_ros_wrapper_node"
+            name="continental_srr520_hw_driver" output="screen" if="$(var launch_hw)">
+            <remap from="/diagnostics" to="diagnostics"/>
+            <remap from="odometry_input" to="$(var odometry_topic)"/>
+            <remap from="acceleration_input" to="$(var acceleration_topic)"/>
+            <remap from="steering_angle_input" to="$(var steering_angle_topic)"/>
+
+            <param name="sensor_model" value="$(var sensor_model)"/>
+            <param name="frame_id" value="$(var frame_id)"/>
+            <param name="base_frame" value="$(var base_frame)"/>
+
+            <param name="interface" value="$(var interface)"/>
+            <param name="receiver_timeout_sec" value="$(var receiver_timeout_sec)"/>
+            <param name="sender_timeout_sec" value="$(var sender_timeout_sec)"/>
+            <param name="filters" value="$(var filters)"/>
+            <param name="use_bus_time" value="$(var use_bus_time)"/>
+
+            <param name="new_longitudinal_cog" value="1.25"/> <!-- test value -->
+            <param name="new_sensor_id" value="$(var new_sensor_id)"/>
+            <param name="new_vehicle_wheelbase" value="$(var new_vehicle_wheelbase)"/>
+            <param name="new_plug_bottom" value="$(var new_plug_bottom)"/>
+            <param name="new_cover_damping" value="$(var new_cover_damping)"/>
+            <param name="reset_sensor_configuration" value="$(var reset_sensor_configuration)"/>
+
+            <param name="use_sim_time" value="false"/>
+        </node>
+    </group>
```
Please move this to #124 if possible :pray:

knzo25 replied:
Addressed in [1007571](https://github.com/tier4/nebula/pull/123/commits/100757164b9eee0ced387ed2c5276687796abbfd)

mojomex commented on nebula_ros/launch/continental_launch_all_hw.xml:
```diff
@@ -0,0 +1,174 @@
+    <arg name="configuration_vehicle_height" default="2.5" description="New vehicle height"/>
+    <arg name="configuration_vehicle_wheelbase" default="2.79" description="New vehicle wheel base"/>
+
+    <!-- CAN based sensor: SRR520 -->
+    <arg name="interface" default="can4" description="Can interface"/>
+    <arg name="receiver_timeout_sec" default="0.03" description="Receiver timeout in seconds"/>
+    <arg name="sender_timeout_sec" default="0.01" description="Sender timeout int seconds"/>
+    <arg name="filters" default="0:0" description="CAN filters"/>
+    <arg name="use_bus_time" default="false" description="Whether to use the bus time for received messages or the system time"/>
+
+    <arg name="new_sensor_id" default="0" description="The new sensor id"/>
+    <arg name="new_plug_bottom" default="true" description="Whether the plug is in the bottom"/>
+    <arg name="new_cover_damping" default="0.0" description="Cover damping in dB"/>
+    <arg name="reset_sensor_configuration" default="false" description="Whether to reset all values to default when configuring the sensor"/>
```
Please move this to #124 if possible :pray:

knzo25 replied:
Addressed in [1007571](https://github.com/tier4/nebula/pull/123/commits/100757164b9eee0ced387ed2c5276687796abbfd)

mojomex commented on nebula_ros/launch/nebula_launch.py:
When launching from the `.py` launch file, how would the vehicle dimension parameters be set? Via a separate `ros2 param set` call?

Radar and LiDAR are quite different regarding their parameters, so I do not know if having all of them in the same launch file (== having all possible parameters in there) is the best approach (@drwnz @amc-nu what do you think?)

knzo25 replied:
Since the config yaml file can be specified, those parameters should be loaded from there I think

```
    with open(sensor_params_fp, "r") as f:
            sensor_params = yaml.safe_load(f)["/**"]["ros__parameters"]
    nodes = []
    launch_hw = LaunchConfiguration("launch_hw").perform(context) == "true"

    if launch_hw:
        nodes.append(
            # HwInterface
            ComposableNode(
                package="nebula_ros",
                plugin=get_plugin_name(sensor_make, sensor_model)+"HwInterfaceRosWrapper",
                name=sensor_make.lower()+"_hw_interface_ros_wrapper_node",
                parameters=[
                    sensor_params,
                    {
                        "sensor_model": LaunchConfiguration("sensor_model"),
                        "sensor_ip": LaunchConfiguration("sensor_ip"),
                        "return_mode": LaunchConfiguration("return_mode"),
                        "calibration_file": calibration_file or sensor_calib_fp,
                        "correction_file": correction_file or sensor_calib_fp,
                        "setup_sensor": LaunchConfiguration("setup_sensor"),
                        "ptp_profile": LaunchConfiguration("ptp_profile"),
                        "ptp_domain": LaunchConfiguration("ptp_domain"),
                        "ptp_transport_type": LaunchConfiguration("ptp_transport_type"),
                        "ptp_switch_type": LaunchConfiguration("ptp_switch_type"),
                    },
                ],
            ),
        )
```

mojomex replied:
Thanks, good catch!

mojomex commented on nebula_ros/src/continental/continental_ars548_hw_interface_ros_wrapper.cpp:
```diff
@@ -0,0 +1,510 @@
+  response->message = (std::stringstream() << result).str();
+}
+
+void ContinentalARS548HwInterfaceRosWrapper::SetVehicleParametersRequestCallback(
+  [[maybe_unused]] const std::shared_ptr<
+    continental_srvs::srv::ContinentalArs548SetVehicleParameters::Request>
+    request,
+  const std::shared_ptr<continental_srvs::srv::ContinentalArs548SetVehicleParameters::Response>
+    response)
+{
+  std::scoped_lock lock(mtx_config_);
+  auto result = hw_interface_.SetVehicleParameters(
+    sensor_configuration_.configuration_vehicle_length,
+    sensor_configuration_.configuration_vehicle_width,
+    sensor_configuration_.configuration_vehicle_height,
+    sensor_configuration_.configuration_vehicle_wheelbase);
+
+  response->success = result == Status::OK;
+  response->message = (std::stringstream() << result).str();
+}
```
It might be unexpected behavior for users to have to send an empty service request to send params to the sensor that were defined in the launch file.

It would be nicer if
* params were sent in the service request, or
* the service is removed and the params in the launch file are set on startup without a service call

knzo25 replied:
In [423112d](https://github.com/tier4/nebula/pull/123/commits/423112d69d8bfec2446adae90cf4ce20dae5c505)
I added an option to set the parameters via the services. Basically:

vehicle parameters: if no parameters are given to the service (`ros2 service call name type {}`), it will use the parameters from the launcher. If parameters are given, it will use those (`ros2 service call name type {parameter_value: overwrite_value}`)

sensor mounting: there is a boolean to autoconfigure the mounting position. if set to true (default) it will use the tfs. If not, it will use the parameters from the service request

mojomex commented on nebula_messages/continental_srvs/srv/ContinentalArs548SetVehicleParameters.srv:
See the comment above about the empty service call

knzo25 replied:
In [423112d](https://github.com/tier4/nebula/pull/123/commits/423112d69d8bfec2446adae90cf4ce20dae5c505)
I added an option to set the parameters via the services. Basically:

vehicle parameters: if no parameters are given to the service (ros2 service call name type {}), it will use the parameters from the launcher. If parameters are given, it will use those (ros2 service call name type {parameter_value: overwrite_value})

sensor mounting: there is a boolean to autoconfigure the mounting position. if set to true (default) it will use the tfs. If not, it will use the parameters from the service request

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,623 @@
+    temperature_status_vector.push_back("Ok");
+  }
+  if (sensor_status_packet.temperature_status & 0x01) {
+    temperature_status_vector.push_back("Current under temperature");
```
(pedantic) undertemperature

knzo25 replied:
Fixed in 
851dd86

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,623 @@
+    temperature_status_vector.push_back("Current under temperature");
+  }
+  if (sensor_status_packet.temperature_status & 0x02) {
+    temperature_status_vector.push_back("Past under temperature");
```
(pedantic) undertemperature

knzo25 replied:
Fixed in  851dd86

knzo25 replied:
Fixed in 
851dd86

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,623 @@
+    temperature_status_vector.push_back("Past under temperature");
+  }
+  if (sensor_status_packet.temperature_status & 0x04) {
+    temperature_status_vector.push_back("Current over temperature");
```
(pedantic) overtemperature

knzo25 replied:
Fixed in  851dd86

knzo25 replied:
Fixed in 
aeaad50

mojomex commented on nebula_decoders/src/nebula_decoders_continental/decoders/continental_ars548_decoder.cpp:
```diff
@@ -0,0 +1,623 @@
+    temperature_status_vector.push_back("Current over temperature");
+  }
+  if (sensor_status_packet.temperature_status & 0x08) {
+    temperature_status_vector.push_back("Past over temperature");
```
(pedantic) overtemperature

knzo25 replied:
Fixed in 
aeaad50

mojomex reviewed (commented):
Except for the few small things noted above, this looks really nice!
I'll test with a PCAP or the real sensor asap for some final checks!

Also thanks for doing proper error detection/handling for all the config and sensor comms :smile:

mojomex commented:
Confirmed working with real hardware (used [this launch file](https://github.com/tier4/aip_launcher/blob/feature/xx1-gen2/aip_xx1_launch/launch/radar.launch.xml with the sensor's IP changed accordingly) while Autoware was running (for odometry etc.).

Tested:
* All topics output data, no obvious errors (NaN, inf, etc.) in output
* All Rviz topics are displayed correctly
* People / retro-reflectors moving in front of the sensor are shown
* Output fields in the pointcloud have no obvious errors
* No errors logged in console

Service calls etc. have not yet been tested because of the open conversation above.

knzo25 commented:
@mojomex 
Applied the requested changes and proposed a new way to set the parameters :pray:

mojomex commented:
:green_circle: Service calls are looking good, too!
This completes both my code review and functionality review, LGTM :green_circle:

As there is still the bug related to mounting angles (the sensor's fault as far as we know), I'm waiting for @knzo25's feedback on whether to merge (with an added disclaimer about angle limits) or to wait.

mojomex commented on build_depends.repos:
```diff
@@ -2,5 +2,5 @@ repositories:
     type: git
-    url: https://github.com/MapIV/transport_drivers
-    version: boost
+    url: https://github.com/knzo25/transport_drivers
```
Please be sure to create a PR against [MapIV/transport_drivers](https://github.com/MapIV/transport_drivers) so we don't have duplicate dependencies across the Autoware stack :pray:

knzo25 replied:
@drwnz 
What happened with replacing the transport drivers?
Should I ask for the PR? Or are we moving away from this?

knzo25 replied:
@mojomex 
Talked with David. I will send the PR to Map IV, but in the meantime we would like to merge this PR

knzo25 replied:
Sent the PR: https://github.com/MapIV/transport_drivers/pull/7

mojomex commented:
> I'm waiting for @knzo25's feedback on whether to merge (with an added disclaimer about angle limits) or to wait.

@knzo25 
If I recall correctly, issues only occur in the object output if the mounting angle of the radar is > 20deg off the longitudinal axis, correct?
Would you mind adding a line to the readme warning users of that? Everything else already looks good :+1:

knzo25 commented:
@mojomex 
I got the thumbs up from David to merge this :)
Regarding the erratic behavior, since nothing is confirmed, I prefer to fix and document it once we hear back from Continental

mojomex reviewed (approved):
LGTM!

mojomex commented:
@amc-nu It looks like all requested changes are addressed, is it okay to go ahead and merge?

amc-nu commented:
Yes, please go ahead!
Thank you @knzo25
